### DISCUSSION TITLE
Is there a way to know the exact number of batches per epoch in espnet2?

### DISCUSSION THREAD

**Discussion by Emrys365** [espnet/espnet#4404] [Q&A]
Hi, I am trying to implement a complicated LR scheduler that increases the LR linearly for some steps (like WarmupLR), and then decreases the LR after each epoch (like StepLR).
The first stage changes the LR after each batch, while the second stage changes the LR after each epoch.

Therefore, I wonder whether it is easy to know the exact number of batches per epoch in espnet2, so that I can implement this kind of LR scheduler as a special instantiation of [AbsBatchStepScheduler](https://github.com/espnet/espnet/blob/master/espnet2/schedulers/abs_scheduler.py#L21).

**Comment by Emrys365** [comment]
I see that the [reporter](https://github.com/espnet/espnet/blob/master/espnet2/train/reporter.py#L130-L132) will count the total number of iterations over all epochs. But it requires training for at least one epoch, which is time-consuming.

Instead, I figured out a not-so-easy way by myself:

  1. modify [espnet2/tasks/abs_task.py](https://github.com/espnet/espnet/blob/master/espnet2/tasks/abs_task.py) as follows:

     <details><summary>Click to expand</summary><div>

     ```diff
     --- a/espnet2/tasks/abs_task.py
     +++ b/espnet2/tasks/abs_task.py
     @@ -447,6 +447,14 @@ class AbsTask(ABC):
                  help='Write the output features from the model when "collect stats" mode',
              )
      
     +        group.add_argument_group("validate_train_iter mode related")
     +        group.add_argument(
     +            "--validate_train_iter",
     +            type=str2bool,
     +            default=False,
     +            help='Perform on "validate_train_iter" mode',
     +        )
     +
              group = parser.add_argument_group("Trainer related")
              group.add_argument(
                  "--max_epoch",
     @@ -1222,6 +1230,42 @@ class AbsTask(ABC):
                      log_interval=args.log_interval,
                      write_collected_feats=args.write_collected_feats,
                  )
     +        elif args.validate_train_iter:
     +            # Perform on validate_train_iter mode. This mode has roles
     +            # - Derive the number of training iterations per epoch
     +            # - Validate the train_iter_factory
     +            logging.info(args)
     +
     +            if torch.distributed.is_available():
     +                from torch.distributed import ReduceOp
     +
     +            if args.multiple_iterator:
     +                train_iter_factory = cls.build_multiple_iter_factory(
     +                    args=args,
     +                    distributed_option=distributed_option,
     +                    mode="train",
     +                )
     +            else:
     +                train_iter_factory = cls.build_iter_factory(
     +                    args=args,
     +                    distributed_option=distributed_option,
     +                    mode="train",
     +                )
     +
     +            iterator = train_iter_factory.build_iter(1)
     +            # [For distributed] Because iteration counts are not always equals between
     +            # processes, send stop-flag to the other processes if iterator is finished
     +            iterator_stop = torch.tensor(0).to("cuda" if args.ngpu > 0 else "cpu")
     +            for iiter, (utt_id, batch) in enumerate(iterator, 1):
     +                if distributed_option.distributed:
     +                    torch.distributed.all_reduce(iterator_stop, ReduceOp.SUM)
     +                    if iterator_stop > 0:
     +                        break
     +            else:
     +                if distributed_option.distributed:
     +                    iterator_stop.fill_(1)
     +                    torch.distributed.all_reduce(iterator_stop, ReduceOp.SUM)
     +            print(f"Total iterations per epoch: {iiter}")
              else:
                  # 6. Loads pre-trained model
                  for p in args.init_param:
     ```

     </div></details>

  2. modify [egs2/TEMPLATE/enh1/enh.sh](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/enh1/enh.sh) as follows:

     <details><summary>Click to expand</summary><div>

     ```diff
     --- a/egs2/TEMPLATE/enh1/enh.sh
     +++ b/egs2/TEMPLATE/enh1/enh.sh
     @@ -38,6 +38,7 @@ inference_nj=32     # The number of parallel jobs in inference.
      gpu_inference=false # Whether to perform gpu inference.
      expdir=exp       # Directory to save experiments.
      python=python3       # Specify python to execute espnet commands
     +validate_train_iter=false # whether to count the number of iterations per epoch instead of starting the training process
      
      # Data preparation related
      local_data_opts= # The options given to local/data.sh.
     @@ -119,7 +120,7 @@ Options:
          --dumpdir       # Directory to dump features (default="${dumpdir}").
          --expdir        # Directory to save experiments (default="${expdir}").
          --python         # Specify python to execute espnet commands (default="${python}").
     -
     +    --validate_train_iter=false   # whether to count the number of iterations per epoch instead of starting the training process
          # Data preparation related
          --local_data_opts # The options given to local/data.sh (default="${local_data_opts}").
      
     @@ -586,14 +587,16 @@ if ! "${skip_train}"; then
                  jobname="${enh_exp}/train.log"
              fi
              # shellcheck disable=SC2086
              ${python} -m espnet2.bin.launch \
                  --cmd "${cuda_cmd} --name ${jobname}" \
                  --log "${enh_exp}"/train.log \
                  --ngpu "${ngpu}" \
                  --num_nodes "${num_nodes}" \
                  --init_file_prefix "${enh_exp}"/.dist_init_ \
                  --multiprocessing_distributed true -- \
                  ${python} -m espnet2.bin.enh_train \
     +                --validate_train_iter ${validate_train_iter} \
                      ${_train_data_param} \
                      ${_valid_data_param} \
                      ${_train_shape_param} \
     ```

     </div></details>

3. run the following command under `egs2/<recipe_name>/enh1`:

     ```bash
     ./run.sh --stage 6 --stop_stage 6 --validate_train_iter true
     ```
